chop/penalties.py

```diff
@@ -0,0 +1,80 @@
+"""
+This module contains various penalties / regularizers.
+They function batch-wise, similar to objects in `chop.constraints`.
+
+Code inspired from https://github.com/openopt/copt/
+"""
+
+import torch
+import torch.nn.functional as F
+
+from chop import utils
+
+
+class L1:
+    """L1 penalty. Batch-wise function."""
+
+    def __init__(self, alpha):
+        if alpha < 0:
+            raise ValueError("alpha must be non negative.")
+        self.alpha = alpha
+
+    def __call__(self, x):
+        batch_size = x.size(0)
+        return abs(x.view(batch_size, -1)).sum(dim=-1)
+
+    def prox(self, x, step_size=None):
+        """Prox operator for L1 norm. This is given by soft-thresholding."""
+        return torch.sign(x) * F.relu(abs(x) - self.alpha * step_size)
+
+
+class GroupL1:
+    """
+    Group LASSO penalty. Batch-wise function.
+    """
+
+    def __init__(self, alpha, groups):
+        self.alpha = alpha
+
+        # TODO: implement ValueErrors
+        # groups must be indices and non overlapping
+        if not isinstance(groups[0], torch.Tensor):
+            groups = [torch.tensor(group) for group in groups]
+        if groups[0].dim() == 0:
+            groups = [torch.tensor([group]) for group in groups]
+        self.groups = groups
+
+    def __call__(self, x):
+        if x.dim() == 2:
+            group_norms = torch.stack([torch.linalg.norm(x[:, g], dim=-1)
+                                       for g in self.groups])
+        else:
+            group_norms = torch.stack([torch.linalg.norm(x[(Ellipsis,) + tuple(zip(*g))],
+                                                         dim=-1)
+                                       for g in self.groups])
+
+        return self.alpha * group_norms.sum(dim=0)
+
+    @torch.no_grad()
+    def prox(self, x, step_size=None):
+        out = x.detach().clone()
+        if type(step_size) == float:
+            step_size *= torch.ones(x.size(0))
+
+        if x.dim() == 2:
```

Comment:
would be nice to have a comment on why this distinction of cases

Reply:
and why it's not present in the L1 prox

Reply:
Right, I think I need to add this to the L1 prox for consistency with the `constraints` module